dev/bots/test.dart

```diff
@@ -241,6 +243,132 @@ Future<Null> _runFlutterAnalyze(String workingDirectory, {
+    .where((FileSystemEntity entity) => entity is File && path.extension(entity.path) == '.dart')
+    .map<String>((FileSystemEntity entity) => path.basenameWithoutExtension(entity.path))
+    .toList()..sort();
+  final List<String> directories = new Directory(srcPath).listSync()
+    .where((FileSystemEntity entity) => entity is Directory)
+    .map<String>((FileSystemEntity entity) => path.basename(entity.path))
+    .toList()..sort();
+  if (!_matches(packages, directories)) {
+    errors.add(
+      'flutter/lib/*.dart does not match flutter/lib/src/*/:\n'
+      'These are the exported packages:\n' +
+      packages.map((String path) => '  lib/$path.dart').join('\n') +
+      'These are the directories:\n' +
+      directories.map((String path) => '  lib/src/$path/').join('\n')
+    );
+  }
+  // Verify that the imports are well-ordered.
+  final Map<String, Set<String>> dependencyMap = new Map<String, Set<String>>.fromIterable(
+    directories,
+    key: (String directory) => directory,
+    value: (String directory) => _findDependencies(path.join(srcPath, directory), errors, checkForMeta: directory != 'foundation'),
+  );
+  for (String package in dependencyMap.keys) {
+    if (dependencyMap[package].contains(package)) {
+      errors.add(
+        'One of the files in the $yellow$package$reset package imports that package recursively.'
+      );
+    }
+  }
+  for (String package in dependencyMap.keys) {
+    final List<String> loop = _deepSearch(dependencyMap, package);
+    if (loop != null) {
+      errors.add(
+        '${yellow}Dependency loop:$reset ' +
+        loop.join(' depends on ')
+      );
+    }
+  }
+  // Fail if any errors
+  if (errors.isNotEmpty) {
+    print('$red━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━$reset');
+    if (errors.length == 1) {
+      print('${bold}An error was detected when looking at import dependencies within the Flutter package:$reset\n');
+    } else {
+      print('${bold}Multiple errors were detected when looking at import dependencies within the Flutter package:$reset\n');
+    }
+    print(errors.join('\n\n'));
+    print('$red━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━$reset\n');
+    exit(1);
+  }
+}
+
+bool _matches<T>(List<T> a, List<T> b) {
+  assert(a != null);
+  assert(b != null);
+  if (a.length != b.length)
+    return false;
+  for (int index = 0; index < a.length; index += 1) {
+    if (a[index] != b[index])
+      return false;
+  }
+  return true;
+}
+
+final RegExp _importPattern = new RegExp(r"import 'package:flutter/([^.]+)\.dart'");
+final RegExp _importMetaPattern = new RegExp(r"import 'package:meta/meta.dart'");
+
+Set<String> _findDependencies(String srcPath, List<String> errors, { bool checkForMeta: false }) {
+  return new Directory(srcPath).listSync().where((FileSystemEntity entity) {
+    return entity is File && path.extension(entity.path) == '.dart';
+  }).map<Set<String>>((FileSystemEntity entity) {
```

Comment:
[error] The function expression type '(File) → Set<String>' isn't of type '(FileSystemEntity) → Set<String>'. This means its parameter or return type does not match what is expected. Consider changing parameter type(s) or the returned type(s).